sdks/python/apache_beam/io/range_trackers_test.py

```diff
@@ -319,6 +320,79 @@ def test_try_test_split_at_position_finite_range(self):
         self.bytes_to_position([3, 2, 1])))
 
 
+class OrderedPositionRangeTrackerTest(unittest.TestCase):
+
+  class DoubleRangeTracker(range_trackers.OrderedPositionRangeTracker):
+
+    @staticmethod
+    def fraction_to_position(fraction, start, end):
+      return start + (end - start) * fraction
+
+    @staticmethod
+    def position_to_fraction(pos, start, end):
+      return float(pos - start) / (end - start)
+
+  def test_try_claim(self):
+    tracker = self.DoubleRangeTracker(10, 20)
+    self.assertTrue(tracker.try_claim(10))
+    self.assertTrue(tracker.try_claim(15))
+    self.assertFalse(tracker.try_claim(20))
+    self.assertFalse(tracker.try_claim(25))
+
+  def test_fraction_consumed(self):
+    tracker = self.DoubleRangeTracker(10, 20)
+    self.assertEqual(0, tracker.fraction_consumed())
+    tracker.try_claim(10)
+    self.assertEqual(0, tracker.fraction_consumed())
+    tracker.try_claim(15)
+    self.assertEqual(.5, tracker.fraction_consumed())
+    tracker.try_claim(17)
+    self.assertEqual(.7, tracker.fraction_consumed())
+    tracker.try_claim(25)
+    self.assertEqual(.7, tracker.fraction_consumed())
+
+  def test_try_split(self):
+    tracker = self.DoubleRangeTracker(10, 20)
+    tracker.try_claim(15)
+    self.assertEqual(.5, tracker.fraction_consumed())
+    # Split at 18.
+    self.assertEqual((18, 0.8), tracker.try_split(18))
+    # Fraction consumed reflects smaller range.
+    self.assertEqual(.625, tracker.fraction_consumed())
+    # We can claim anything less than 18,
+    self.assertTrue(tracker.try_claim(17))
+    # but can't split before claimed 17,
+    self.assertIsNone(tracker.try_split(16))
+    # nor claim anything after 18.
+    self.assertFalse(tracker.try_claim(19))
+
+  def test_claim_order(self):
+    tracker = self.DoubleRangeTracker(10, 20)
+    tracker.try_claim(12)
+    tracker.try_claim(15)
+    with self.assertRaises(ValueError):
+      tracker.try_claim(13)
+
+  def test_out_of_range(self):
+    tracker = self.DoubleRangeTracker(10, 20)
+    # Can't claim before range.
+    with self.assertRaises(ValueError):
+      tracker.try_claim(-5)
+    # Can't split before range.
+    with self.assertRaises(ValueError):
+      tracker.try_split(-5)
+    # Can't split at start position.
```

Comment:
I'm not sure this should be an error - perhaps it should just return false. Logically, it produces an empty range and the full range, which is degenerate and useless (which is why we should reject it) but mathematically sound (which is why we shouldn't count it as an error).